srml/randomness/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Randomness Module
+//!
+//! The Randomness module provides a [`random`](./struct.Module.html#method.random) function that
+//! generates low-influence random values based on the block hashes from the previous 81 blocks.
+//!
+//! ## Public Functions
+//!
+//! See the [`Module`](./struct.Module.html) struct for details of publicly available functions.
+//!
+//! ## Usage
+//!
+//! ### Prerequisites
+//!
+//! Import the Randomness module and derive your module's configuration trait from the system trait.
+//!
+//! ### Example - Get random seed for the current block
+//!
+//! ```
+//! use support::{decl_module, dispatch::Result};
+//!
+//! trait Trait: system::Trait {}
+//!
+//! decl_module! {
+//! 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//! 		pub fn random_module_example(origin) -> Result {
+//! 			let _random_seed = <srml_random::Module<T>>::random_seed();
+//! 			Ok(())
+//! 		}
+//! 	}
+//! }
+//! # fn main() { }
+//! ```
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use rstd::prelude::*;
+use sr_primitives::traits::Hash;
+use support::{decl_module, decl_storage};
+use safe_mix::TripletMix;
+use codec::Encode;
+use system::Trait;
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn on_initialize() {
+			let parent_hash = <system::Module<T>>::parent_hash();
+
+			<RandomMaterial<T>>::mutate(|&mut(ref mut index, ref mut values)| if values.len() < 81 {
+				values.push(parent_hash)
+			} else {
+				values[*index as usize] = parent_hash;
+				*index = (*index + 1) % 81;
```

Comment:
I know that you just copied it, but isn't the `index == block_number % 81`?

Reply:
Yep, I'll change that.

Reply:
actually, it's just either to leave it as-is:
```
error[E0308]: mismatched types
  --> srml/randomness/src/lib.rs:68:29
   |
68 |                 *index = block_number % 81;
   |                                         ^^ expected associated type, found integer
   |
   = note: expected type `<T as srml_system::Trait>::BlockNumber`
              found type `{integer}`
```

Reply:
`81.into()`

Reply:
And I meant to not store this index anymore.

Reply:
Oh ok, that works.

Reply:
Hmm, `T::BlockNumber` implementing `TryInto<usize>` but not `Into<usize>` is a bit of a problem if we're trying to avoid panics. I could use `unique_saturating_into` though which casts to `u128` and then cast that to `usize`. What do you suggest?

Reply:
Just use expect. You can prove it statically that `something % 81` always generates a number that is between `0..81` and that is way below `usize::max_value()`.